### ISSUE TITLE
Supporting typed errors on the HttpErrorHandler

### ISSUE THREAD

**Issue by AlexITC** [playframework/playframework#8459]
At the moment, the [HttpErrorHandler#onClientError](https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/http/HttpErrorHandler.scala#L37) supports a string message only.

It would be useful to be able to use custom errors instead (like typed errors).

Some use cases could be allowing to return a custom error while using the `QueryStringBindable` or the `PathBindable`, at the moment, we are required to return a `String` for the error case (see [QueryStringBindable#bind](https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/mvc/Binders.scala#L80) and [PathBindable#bind](https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/mvc/Binders.scala#L165)).

I understand it is difficult to perform the required changes due to the need to redesign several classes, the simplest but ugly solution that I see is to change the `String` from the `HttpErrorHandler` to  `AnyRef`.

In case this is something interesting for you, I could invest some time to provide a design or a pull request, what do you think?

**Comment by marcospereira** [comment]
Hi @AlexITC,

Do you want to have something like:

```scala
def onClientError(request: RequestHeader, statusCode: Int, error: Error)
```
And then:

```scala
trait PathBindable[A] {
  def bind(key: String, value: String): Either[PathBindableError, A]
}
```
Where `PathBindableError extends Error`. Is that what you are proposing? 

It should be possible to maintain both APIs at the same time, without the need to change from `String` to `AnyRef` in `HttpErrorHandler`. And if we decide it is worth, then we can deprecate the `String` version and remove it later.

But I'm unsure about what value this change brings. What will the users do differently (and better) from what exists now?

**Comment by AlexITC** [comment]
Yes, something like that should work, the advantage to me is to simplify how to keep all responses consistent in my applications, having this I will be abla to use the `PathBindable`.

Having a different API is ok.

**Comment by mkurz** [comment]
@AlexITC will you work on this?

**Comment by AlexITC** [comment]
@mkurz I would like to, sadly, my time has been quite limited and I expect to continue in this way for a couple of months.

**Comment by fusuiyi123** [comment]
may I work on this:)
